### ISSUE TITLE
use torch.backends.cudnn.deterministic = True causes slow running speed

### ISSUE THREAD

**Issue by yichenwang231** [oezyurty/CLUDA#5]
Thank you for your work.

When I use ‘torch.backends.cudnn.deterministic = True’ to fix the random seed to ensure that the results of each run are the same, the running speed of the model is dozens of times slower, this issue has not appeared in other frameworks. How can I solve this problem in your framework. thank you.

**Comment by oezyurty** [comment]
Thanks for your interest in our work! 

To start with a short and honest answer, I don't know the exact reason as I don't have much experience with torch.backends.cudnn.deterministic . 

That's why I am just throwing some ideas below, maybe such brainstorming helps locating the root cause.
- You wrote that this issue has not appeared in other frameworks. Do you mean the other baseline algorithms we compare? If that's the case, then temporal convolution network (TCN) cannot be the reason, as we use the same TCN architecture across baselines and our CLUDA. (But if you meant completely different architecture, then I would first start investigating TCN, as I assume cudnn as some tailored approximations for convolutional kernels)
- If we can eliminate that TCN is not the reason of slow down, I believe our most compute-intensive part comes from MoCo contrastive learning framework. If you use a large queue (e.g. ~80-90K as we originally used), it may result in large matrix multiplications. If cudnn had some approximations to improve the runtime, then this deterministic version may cause the slow-down. (albeit sacrificing the accuracy, maybe you may try very low queue sizes to see if that reduces the runtime. Note that (as far as I remember) queue size still needs to be k time batch_size, e.g. k is any integer.)

As our original framework does not involve any stochasticity (besides random initialization), it seems like a matter of locating where cudnn automatically opts approximation. Sorry that I cannot pinpoint the source for you. Best,

**Comment by yichenwang231** [comment]
Thank you very much!
It is indeed the reason for TCN，[init_weights](https://github.com/oezyurty/CLUDA/blob/7bad2deb4c975afdec43bebc6eeb329ff5ba9081/utils/tcn_no_norm.py#L37),when I use `from torch.nn.utils import weight_norm` to init self.conv1,self.conv2 and self.downsample,the running speed will be normal,I don't think there is much difference between these two init methods,do you have a different perspective?
Thank you again!Best.
